<a href="https://colab.research.google.com/github/m-nganga/Colab-Notebooks/blob/master/SQL_Basics_for_Data_Science_Project_(24th_June_2020).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SQL Basics for Data Science: Workshop Project (24th June 2020)

## Project Deliverable 


* Project solution crafted after the analysis of the business situation herebelow.

You are a Data Science Consultant working on a project to understand car crashes in the city of Chicago. You are required to make recommendations to the office of the mayor of Chicago on the steps to take that would lead to a reduction in car crashes, based on the findings from the dataset analysis findings . 


As part of your descriptive analysis report, you decide to find answers to the following questions using SQL.
1. Display a list of 5 makes that had the highest number of crashes?
2. Most personal vehicles involved in crashes were travelling towards which direction? 
3. How many passengers were involved in personal cars were in accidents?
4. How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?
5. Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.
6. Compute the total occupant count per vehicle count.
7. Sort the number of passengers in descending order by travel direction taking into account travel direction. Return travel direction and no. of passengers.
8. Select top 5 vehicle models involved in the accident with no defect, order them by make and show their manoeuvre. 
9. Which ridesharing vehicles in an accident had the most number of passengers? Return model, and no of passengers.
 
<strong>Remember:</strong> What matters the most is what your insights mean to the business. Make sure to have a section where you elaborate how your analysis findings tie to your recommendations.
 


## Dataset Explained


Before performing any analysis, you should select only vehicles manufactured in the last 15 years from MySQL traffic_accidents_database whose URL and credentials are given below.



URL = http://***/<br>
username = *** <br>
password = ***


## Preparing Data

After connecting to the database using the credentials provided, I have selected the required set of records by running the following SQL query.

SELECT * FROM `traffic_accidents` WHERE vehicle_year > 2005  AND vehicle_year <= 2020

<em> <font color = lightblue>I am sure we can make use of the various date() functions here to select the years between now and 15 years ago, but **not sure of the syntax for now**. currently researching on this. </font></em>

I have thereafter exported the resulting recordset as a csv file for downstream analysis through this notebook.

## Step 1. Importing Required Libraries

In [2]:
# Importing pandas
import pandas as pd

# Loading SQL extension
%load_ext sql

# Connecting sqlite database
%sql sqlite://

'Connected: @None'

## Step 2. Importing Dataset

In [3]:
# Load the dataset from a csv file and store it in a dataframe called traffic_accidents.
t_accidents = pd.read_csv('traffic_accidents.csv')


#Store the dataset in our in memory sqlite database. but for controls, check 
#first if the table exists in the database, and if so drop it. 
%sql DROP TABLE if EXISTS retail;


#Finally create an SQL table of our sqlite database running in memory and store 
#the recordset in readiness for further analysis with SQL 
%sql PERSIST t_accidents; 


 * sqlite://
Done.
 * sqlite://


'Persisted t_accidents'

### 2.1 Sample Recordset

In [4]:
# Sample the first few records of our dataset to understand the structure of the dataset
%%sql 
SELECT * FROM t_accidents LIMIT 10

 * sqlite://
Done.


[(0, 758252, 'JC469272', '2019-10-11T19:45:00', 1, 'DRIVER', None, 720258, None, 'DODGE', 'CARAVAN', 'IL', 2006, 'UNKNOWN', 'VAN/MINI-VAN', 'PERSONAL', 'W', 'STRAIGHT AHEAD', 1, None, 1, None, 'E&R TOWING', '10301 S DOTY RD', None, None, None, None, None, None, 0, None, None, None, None, None, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (1, 758164, 'JC469222', '2019-10-11T18:39:00', 1, 'DRIVER', None, 720177, None, 'CHEVROLET', 'COBALT', 'CA', 2006, 'UNKNOWN', 'PASSENGER', 'UNKNOWN/NA', 'UNKNOWN', 'CHANGING LANES', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1, 'TOTAL (ALL AREAS)', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (2, 758273, 'JC469228', '2019-10-11T18:29:00', 1, 'DRIVER', None, 720287, None, 'KIA MOTORS CORP', 'Optima', 'IL', 2006, 'NONE', 'PASSENGER', 'PERSONAL', 'E', 'TURNING LEFT', None, None, 1, None, None, None, None, 1, None, None, None, None, None, None, None, None, None, 1, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (3, 758074, 'JC469085', '2019-10-11T17:10:00', 1, 'DRIVER', None, 720081, None, 'MERCEDES-BENZ', 'UNKNOWN', 'IL', 2006, 'UNKNOWN', 'PASSENGER', 'UNKNOWN/NA', 'E', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (4, 758005, 'JC469056', '2019-10-11T16:30:00', 1, 'DRIVER', None, 720025, None, 'TOYOTA MOTOR COMPANY, LTD.', 'Matrix (sport wagon)', 'IL', 2006, 'UNKNOWN', 'PASSENGER', 'PERSONAL', 'W', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'UNDER CARRIAGE', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (5, 758006, 'JC469056', '2019-10-11T16:30:00', 2, 'DRIVER', None, 720034, None, 'HONDA', 'Odyssey (minivan)', 'IL', 2006, 'UNKNOWN', 'VAN/MINI-VAN', 'PERSONAL', 'W', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 'SIDE-RIGHT', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (6, 757934, 'JC468930', '2019-10-11T15:50:00', 1, 'DRIVER', 2.0, 719946, None, 'JEEP', 'Liberty', 'IL', 2006, 'NONE', 'PASSENGER', 'PERSONAL', 'N', 'STRAIGHT AHEAD', None, None, 3, None, None, None, None, 1, None, None, None, None, None, None, None, None, None, 1, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None),
 (7, 757929, 'JC468890', '2019-10-11T15:25:00', 1, 'DRIVER', None, 719947, None, 'FORD', 'F150', 'IL', 2006, 'UNKNOWN', 'PASSENGER', 'PERSONAL', 'W', 'STRAIGHT AHEAD', None, None, 1, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, 1, None, 'OTHER', None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, None, No

## Step 3. Business Needs

**1. Which are the top 5 car makes that had the highest number of crashes in the study period?**

In [5]:
#We make use of the property 'make' and use the COUNT()function against this 
#variable for each of its make group. 

%%sql 
SELECT make as 'car make', count(make) as 'accident frequency'
FROM t_accidents
GROUP BY make
ORDER BY count(make) desc
LIMIT 5;

 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 609),
 ('FORD', 545),
 ('CHEVROLET', 538),
 ('NISSAN', 478),
 ('HONDA', 345)]

<em><font color = lightblue>'TOYOTA MOTOR COMPANY, LTD' with 609 accidents had the highest number of accidents reported for the period betweeen 2005 and 2020, followed by FORD with 545, CHEVROLET with 538, NISSAN with 478 and HONDA at 5th position with 345 accidents reported.</font> </em>



---



**2. Towards which direction were the most personal vehicles involved in crashes travelling?**

In [6]:
# We use 'vehicle_type' and 'travel_dierction' to respond to the need above. 
# further we use COUNT() funtion for each of the directions to determine the mode of direction.
%%sql
SELECT travel_direction as 'travel direction', count(travel_direction) as 'crash frequency'
FROM t_accidents
WHERE vehicle_use = 'PERSONAL'
GROUP BY travel_direction 
ORDER BY count(travel_direction) desc
LIMIT 5;

 * sqlite://
Done.


[('N', 849), ('S', 807), ('W', 765), ('E', 708), ('UNKNOWN', 116)]

<em><font color = lightblue>The highest number of passenger cars, at 849, reported with accidents were travelling towards the East</font></em>



---



**3. How many passengers were involved in personal cars accidents??**

In [26]:
# Use the SUM() function to get the total number of passengers for all the 
# personal cars involved in accidents.
%%sql
SELECT SUM(num_passengers) as 'No of Passengers'
FROM t_accidents
WHERE vehicle_use = 'PERSONAL'

 * sqlite://
Done.


[(827.0,)]

<em><font color = lightblue>A total of 827 passengers were involved in personal car accidents.</font></em>



---



**4. How many passengers were involved in accidents while travelling in personal Sport Utility vehicles?**

In [27]:
# Use the SUM() function to get the total number of passengers for all the 
# personal sport utility vehicles involved in accidents.
%%sql
SELECT SUM(num_passengers) as 'No of Passengers'
FROM t_accidents
WHERE vehicle_use = 'PERSONAL' AND vehicle_type = 'SPORT UTILITY VEHICLE (SUV)'


 * sqlite://
Done.


[(137.0,)]

<em><font color = lightblue>A total of 137 passengers were involved in personal SUV car accidents.</font></em>



---



**5. Compute the total number of Mercedez-Benz vehicles involved in accidents while being parked.**

In [9]:
#Use COUNT() function to get the total number of Mercedez-Benz cars.

%%sql
SELECT count(make) as 'No of Mercedez-Benz'
FROM t_accidents
WHERE make = 'MERCEDES-BENZ' AND maneuver = 'PARKED'

 * sqlite://
Done.


[(10,)]

<em><font color = lightblue>Only 10 Mercedez-Benz cars were involved in an accident while parked.</font></em>



---



**6.Compute the total occupancy count per vehicle count.**

In [28]:
# I have interpreted "per vehicle count" to mean for each make of the vehicle.
# Use the SUM() function to get the total number of passengers for each of the makes of the vehicles.
%%sql
SELECT make, SUM(occupant_cnt) as 'Occupancy'
FROM t_accidents
GROUP BY make

 * sqlite://
Done.


[('(UNLISTED MANUFACTURER)', 1),
 ('ACURA', 34),
 ('ACURA (DIV. OF AMERICAN HONDA MOTOR CO.)', 15),
 ('ALFA ROMEO', 0),
 ('AM GENERAL/HUMMER', 3),
 ('AMERICAN TRAILER & MFG. CO. (SEE ALSO TRANSPORTER)', 1),
 ('AMTRAN CORP, GENSIS BUSES', 7),
 ('ASTON-MARTIN', 1),
 ('AUDI', 60),
 ('AUTOCAR', 6),
 ('BANTAM CRANE', 1),
 ('BENTLEY', 0),
 ('BLUE BIRD BODY CO.', 7),
 ('BLUEBIRD INTERNATIONAL', 13),
 ('BMW', 55),
 ('BUICK', 88),
 ('CADILLAC', 49),
 ('CALDWELL DIV., SMITH INTERNATIONAL, INC.', 1),
 ('CASCADE COACH CO.', 1),
 ('CHEVELLE MOBILE HOMES, INC.', 1),
 ('CHEVROLET', 590),
 ('CHRYSLER', 135),
 ('CONTINENTAL', 1),
 ('DODGE', 280),
 ('EIDAL INTERNATIONAL CORP.', 1),
 ('EL DORADO INTERNATIONAL, INC.; EL DORADO MFG., LTD.', 3),
 ('FERRARI', 0),
 ('FIAT', 5),
 ('FIESTA (IMPORTED BY FORD)', 3),
 ("FLEET CAP'N TRAILERS", 4),
 ('FLEETCRAFT CORP.', 1),
 ('FLEETWOOD ENTERPRISES', 1),
 ('FONTAINE TRUCK EQUIPMENT CO.', 1),
 ('FORD', 627),
 ('FREIGHTLINER CORP.', 49),
 ('GENERAL MOTORS CORP.', 74),
 ('GILL MFG. CO.', 3),
 ('GOODYEAR CAMPER MFG., INC., LINCOLN, NB', 1),
 ('GREAT DANE TRAILERS INC.', 1),
 ('HARLEY-DAVIDSON', 3),
 ('HAULER-TRAILER(MFD. BY NORDINE MFG.)', 1),
 ('HINDUSTAN', 1),
 ('HINO', 5),
 ('HONDA', 364),
 ('HYUNDAI', 303),
 ('ICL CARGO TRAILER', 15),
 ('INFINITI', 56),
 ('INTENATIONAL INDUSTRIES, INC.', 3),
 ('INTERNATIONAL COACH MFG.', 5),
 ('INTERNATIONAL HARVESTER', 4),
 ('INTERNATIONAL TRAILER CORP.', 9),
 ('ISUZU', 11),
 ('JAGUAR', 5),
 ('JEEP', 255),
 ('KAWASAKI MOTORS CORP.', 4),
 ('KENWORTH MOTOR TRUCK CO.', 10),
 ('KIA MOTORS CORP', 176),
 ('LAND ROVER', 12),
 ('LEXUS', 73),
 ('LINCOLN ELECTRIC CO., INC. CLEVELAND, OH ; LINCOLN WELDER', 1),
 ('LINCOLN-CONTINENTAL', 53),
 ('MACK TRAILER MFG. CO., INC., DETRIOT, MI', 1),
 ('MACK TRUCKS, INC.', 6),
 ('MAZDA', 81),
 ('MERCEDES-BENZ', 64),
 ('MERCURY', 7),
 ('MINI', 13),
 ('MITSUBISHI', 32),
 ('MOTOR COACH INDUSTRIES, INC.', 2),
 ('NAVISTAR', 1),
 ('NEW FLYER', 57),
 ('NISSAN', 553),
 ('NISSAN DIESEL MOTOR CO.', 1),
 ('OLDSMOBILE', 1),
 ('PETERBILT MOTORS CO., DIVISION PACCAR, INC.', 6),
 ('PONTIAC', 46),
 ('PORSCHE', 8),
 ('PREVOST CAR, BUS, SAINTE CLAIRE, QUEBEC', 12),
 ('QUALITY TRAILER', 0),
 ('RANGE ROVER OF NORTH AMERICA', 3),
 ('ROLLS INTERNATIONAL TRAILER(MOBILE HOME)', 1),
 ('SAAB', 3),
 ('SATURN', 19),
 ('SCION', 17),
 ('SCOOTALONG', 1),
 ('STERLINE FLATBED TRAILER', 1),
 ('STERLING', 1),
 ('SUBARU', 70),
 ('SUZUKI', 8),
 ('TESLA MOTORS', 5),
 ('THOMAS BUILT BUS CO.', 1),
 ('TOWMASTER', 0),
 ('TOYOCAR VAN CONTAINER TRAILER', 1),
 ('TOYOTA MOTOR COMPANY, LTD.', 669),
 ('TRIUMPH', 0),
 ('UNKNOWN', 92),
 ('VOLKSWAGEN', 109),
 ('VOLVO', 29),
 ('WABCO CONSTRUCTION & MINING EQUIPMENT GROUP', 2),
 ('WORK HORSE MFG.', 1),
 ('YAMAHA', 1)]

*I have interpreted **"per vehicle count"** to mean for each make of the vehicle*

 

---



**7. Sort the number of passengers in descending order by travel direction taking into account travel direction.**

In [30]:
# We will need data from  'occupant_cnt', travel_direction, and 'travel_direction' 
# to respond to the need above. 
# further we use COUNT() function on occupant_cnt to get the number of passengers 
# for each of the travel directions
%%sql
SELECT travel_direction as 'Travel Direction', count(num_passengers) as 'No. of Passengers'
FROM t_accidents
GROUP BY travel_direction 
ORDER BY count(num_passengers) desc, travel_direction

 * sqlite://
Done.


[('N', 198),
 ('W', 194),
 ('E', 184),
 ('S', 184),
 ('SE', 12),
 ('UNKNOWN', 9),
 ('NE', 6),
 ('NW', 6),
 ('SW', 6)]

<em><font color = lightblue>Travel direction `NORTH` recorded the highest number of passengers involved in accidents with `198` people.</font></em>



---



**8. Select top 5 vehicle models involved in the accident with no defect, order them by make and show their manoeuvre.**

In [25]:
# We use 'model' and 'vehicle_defect', 'make' and maneuver to respond to the need above
# but pick only cases where vehicles didnt have any defects. 
# further we use ORDER BY Clause for sorting by 'make' and finally LIMIT as we are only interested in the top 5
%%sql
SELECT make, model, maneuver, count(model) as 'vehicle model'
FROM t_accidents
WHERE vehicle_defect = 'NONE'
GROUP BY make, model
ORDER BY count(model) DESC, make ASC
LIMIT 5

 * sqlite://
Done.


[('TOYOTA MOTOR COMPANY, LTD.', 'CAMRY', 'STRAIGHT AHEAD', 129),
 ('NISSAN', 'NISSAN ALTIMA', 'PARKED', 78),
 ('NISSAN', 'SENTRA (DATSUN AND NISSAN HAVE MERGED)', 'SLOW/STOP IN TRAFFIC', 71),
 ('CHEVROLET', 'MALIBU (CHEVELLE)', 'PARKED', 68),
 ('FORD', 'UNKNOWN', 'STRAIGHT AHEAD', 60)]

<em><font color = lightblue>From the analysis, `Toyota Camry` vehicles which had no defects registered the highest number of accidents with `129` of them.</font></em>



---



**9. Which ridesharing vehicle models involved accidents had the most number of passengers? Return model, and no of passengers?**

In [31]:
# We use 'model', 'num_passengers' and 'vehicle_use' to pick relevant records. 
# further we use COUNT() funtion for each of the directions to determine the mode of direction.
%%sql
SELECT model, count(num_passengers) as 'no. of passengers'
FROM t_accidents
WHERE vehicle_use = 'RIDESHARE SERVICE'
GROUP BY model 
ORDER BY count(num_passengers) desc
LIMIT 5;

 * sqlite://
Done.


[('UNKNOWN', 4),
 ('CAMRY', 3),
 ('COROLLA', 3),
 ('NISSAN ALTIMA', 3),
 ('FUSION', 2)]

<em><font color = lightblue> The ridesharing model of vehicles which had most number of passengers were, unfortunately, not classified but registered only as "Unknown"</font></em>

## Step 4. Recommendations

From the foregoing analysis, we are able to make the following recommendations to the Office of Mayor of Chicago on the steps to take that would lead to a reduction in car crashes. 

* **Recommendation:** Carry out deeper root-cause analysis (RCA) on the following; 


1.   Why Toyota and Ford make of vehicles are the most frequently make of   cars involved in crashes. (Could it be because of the fact that there are more of such cars on the road than other makes hence more probable to make crashes?)

2.   Why cars travelling to the North and to the South are the most frequently cars involved in crashes. (Could it be the state of of the North-South Road?, other associated road users along the road like many pedestrians or large haulier trucks? etc.) 
